lib/internal/vm/Module.js

```diff
@@ -0,0 +1,205 @@
+    if (url !== undefined) {
+      if (typeof url !== 'string') {
+        throw new errors.TypeError(
+          'ERR_INVALID_ARG_TYPE', 'options.url', 'string', url);
+      }
+      url = new URL(url).href;
+    } else if (context === undefined) {
+      url = `vm:module(${globalModuleId++})`;
+    } else if (perContextModuleId.has(context)) {
+      const curId = perContextModuleId.get(context);
+      url = `vm:module(${curId})`;
+      perContextModuleId.set(context, curId + 1);
+    } else {
+      url = 'vm:module(0)';
+      perContextModuleId.set(context, 1);
+    }
+
+    const wrap = new ModuleWrap(src, url, {
+      [kParsingContext]: context,
+      lineOffset: options.lineOffset,
+      columnOffset: options.columnOffset
+    });
+
+    wrapMap.set(this, wrap);
+    linkingStatusMap.set(this, 'unlinked');
+
+    Object.defineProperties(this, {
+      url: { value: url, enumerable: true },
+      context: { value: context, enumerable: true },
+    });
+  }
+
+  get linkingStatus() {
+    return linkingStatusMap.get(this);
+  }
+
+  get status() {
+    return STATUS_MAP[wrapMap.get(this).getStatus()];
+  }
+
+  get namespace() {
+    const wrap = wrapMap.get(this);
+    if (wrap.getStatus() < kInstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS',
+                             'must not be uninstantiated or instantiating');
+    return wrap.namespace();
+  }
+
+  get dependencySpecifiers() {
+    let deps = dependencyCacheMap.get(this);
+    if (deps !== undefined)
+      return deps;
+
+    deps = wrapMap.get(this).getStaticDependencySpecifiers();
+    Object.freeze(deps);
+    dependencyCacheMap.set(this, deps);
+    return deps;
+  }
+
+  get error() {
+    const wrap = wrapMap.get(this);
+    if (wrap.getStatus() !== kErrored)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be errored');
+    return wrap.getError();
+  }
+
+  async link(linker) {
+    if (typeof linker !== 'function')
+      throw new errors.TypeError(
+        'ERR_INVALID_ARG_TYPE', 'linker', 'function', linker);
+    if (linkingStatusMap.get(this) !== 'unlinked')
+      throw new errors.Error('ERR_VM_MODULE_ALREADY_LINKED');
+    const wrap = wrapMap.get(this);
+    if (wrap.getStatus() !== kUninstantiated)
+      throw new errors.Error('ERR_VM_MODULE_STATUS', 'must be uninstantiated');
+    linkingStatusMap.set(this, 'linking');
+    const promises = [];
+    wrap.link((specifier) => {
+      const p = (async () => {
+        const m = await linker(this, specifier);
+        if (!m || !wrapMap.has(m))
+          throw new errors.Error('ERR_VM_MODULE_NOT_MODULE');
+        if (m.context !== this.context)
+          throw new errors.Error('ERR_VM_MODULE_DIFFERENT_CONTEXT');
+        const childLinkingStatus = linkingStatusMap.get(m);
+        if (childLinkingStatus === 'errored')
+          throw new errors.Error('ERR_VM_MODULE_LINKING_ERRORED');
+        if (childLinkingStatus === 'unlinked')
+          await m.link(linker);
```

Comment:
this is a bit of an awkward thing to do, the user might intend to link that module in a different way since the link method is specific to each module.  (although that might not be a wise thing to do on their part)

Reply:
If they does, they may link it before returning from the asynchronous linker function, in which case `childLinkingStatus` will either be `'linking'` or `'linked'`.